File: 백준/35274_제설_작업_2/main.kt
import java.io.BufferedReader
import java.io.BufferedWriter
import java.io.InputStreamReader
import java.io.OutputStreamWriter

fun main(args: Array<String>) {
    val reader = BufferedReader(InputStreamReader(System.`in`))
    val writer = BufferedWriter(OutputStreamWriter(System.`out`))

    class Solution {

        private var N: Int
        private var M: Long
        private var L: Int
        private lateinit var A: LongArray
        private final val limit = 1_000_000_000_000_001

        init {
            val tokens = reader.readLine().split(" ")
            N = tokens[0].toInt()
            M = tokens[1].toLong()
            L = tokens[2].toInt()
            A = reader.readLine()
                .split(" ")
                .map { it -> it.toLong() }
                .toLongArray()
        }

        fun check(X: Long, diff: LongArray): Boolean {
            diff.fill(0L)
            var usedCount = 0L

            var i = 0
            while (i < N) {
                val snow = A[i] + diff[i]

                if (snow <= 0L) {
                    i++
                    continue
                }

                val fullTurns = snow / X
                val remainder = snow % X

                if (fullTurns > 0) {
                    usedCount += fullTurns
                    if (usedCount > M) return false
                    diff[i] -= fullTurns * X
                }

                if (remainder == 0L) {
                    i++
                    continue
                }

                // remainder > 0: i 나머지 제거 + 남은 예산으로 i+1부터 처리
                var remaining = X - remainder
                diff[i] -= remainder
                usedCount++
                if (usedCount > M) return false

                for (j in i + 1 until minOf(N, i + L)) {
                    val sj = A[j] + diff[j]
                    if (sj <= 0L) continue
                    val removed = minOf(sj, remaining)
                    diff[j] -= removed
                    remaining -= removed
                    if (remaining == 0L) break
                }

                i++
            }
            return true
        }

        fun solve(): Long {
            var lo = 0L
            var hi = A.sum() + 1L
            val diff = LongArray(N + 1)
            while (lo + 1L < hi) {
                val mid = lo + (hi - lo) / 2L
                if (!check(mid, diff)) {
                    lo = mid
                } else {
                    hi = mid
                }
            }
            return if (hi > 0 && check(hi, diff)) hi else -1
        }

        fun run() {
            writer.write("${solve()}\n")
            writer.flush()
        }
    }

    val solution = Solution()
    solution.run()
}

File: 백준/35274_제설_작업_2/main.rs
fn check(N: i64, M: i64, L:i64, x: i64, A: &Vec<i64>, diff: &mut Vec<i64>) -> bool {
    diff.fill(0);
    let mut used: i64 = 0; // 사용횟수 
    let mut i: usize = 0;

    while i < N as usize {
        let curr = A[i] + diff[i];

        if curr <= 0 {
            i+=1;
            continue;
        }
        let quot: i64 = curr / x;
        let remainder: i64 = curr % x;

        if quot > 0 {
            used += quot ;
            if used > M {
                return false;
            } 
            diff[i] -= (quot * x);
        }

        if remainder == 0 {
            i+=1;
            continue;
        }

        let mut remaining: i64 = x - remainder;
        diff[i] -= remainder;
        used+=1;

        if used > M  {
            return false; 
        }
        
            // j = [i+1, ...] 에 대해 remainder 를 diff에 누적시킬것 
        let limit: usize = std::cmp::min(N as usize, i+L as usize);
        for j in i+1..limit as usize {
            let j_curr = A[j] + diff[j];

            if j_curr <= 0 {
                continue;
            }

            let removed:i64 = std::cmp::min( j_curr, remaining );
            diff[j]-=removed;
            remaining -= removed;

            if remaining == 0 {
                break;
            }
        }

        i+=1;
    }

    true
}

fn solve(N: i64, M: i64, L: i64) -> i64 {
    let A = io::read_line()
        .split_whitespace()
        .map(|x| x.parse::<i64>().unwrap())
        .collect::<Vec<i64>>();

    let mut lo: i64 = 0;
    let mut hi: i64 = A.iter().sum::<i64>() + 1 as i64;
    let mut diff: Vec<i64> = vec![0; (N+1) as usize];

    while lo + 1 < hi {
        let mid: i64 = lo + (hi - lo) / 2;
        // println!("mid : {}", mid);

        if !check(N,M,L, mid, &A, &mut diff) {
            lo = mid
        } else {
            hi = mid
        }
    }

    if check(N, M, L, hi, &A, &mut diff) { hi } else {-1}
}

fn main() {
    let tok = io::read_line()
        .split_whitespace()
        .map(|x| x.parse::<i64>().unwrap())
        .collect::<Vec<i64>>();
    // for t in tok {
        // println!("t : {}",t);
    // }
    // println!("show inputs");
    // println!("{} {} {}", tok[0], tok[1], tok[2]);
    println!("{}", solve(tok[0], tok[1], tok[2]));
}

mod io {
    pub fn read_line() -> String {
        let mut buf = String::new();
        std::io::stdin().read_line(&mut buf).unwrap();
        buf.trim().to_string()
    }
}
